src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,471 @@
+            }
+            // we should be able to combine these Lists of keys into single SecretKey Objects,
+            // unless we were passed something bogus or an unexportable P11 key
+            try {
+                this.initialKeyMaterial = consolidateKeyMaterial(ikms);
+                this.salt = consolidateKeyMaterial(salts);
+            } catch (InvalidKeyException ike) {
+                throw new InvalidParameterSpecException(
+                    "Issue encountered when combining ikm or salt values into single keys");
+            }
+            // set this value in the "if"
+            if ((info = anExtractExpand.info()) == null) {
+                info = new byte[0];
+            }
+            // set this value in the "if"
+            if ((length = anExtractExpand.length()) <= 0) {
+                throw new InvalidParameterSpecException("length cannot be <= 0");
+            }
+            HKDFTYPE = HKDFTYPES.EXTRACTEXPAND;
+        } else {
+            throw new InvalidParameterSpecException(
+                "The KDFParameterSpec object was not of a recognized type");
+        }
+    }
+
+    private static boolean isNullOrEmpty(Collection<?> c) {
+        return c == null || c.isEmpty();
+    }
+
+    private SecretKey consolidateKeyMaterial(List<SecretKey> keys) throws InvalidKeyException {
+        if (keys != null && !keys.isEmpty()) {
+            ArrayList<SecretKey> localKeys = new ArrayList<>(keys);
+            if (localKeys.size() == 1) {
+                // return this element
+                return localKeys.get(0);
+            } else {
+                byte[] bb = new byte[0];
+                for (SecretKey workItem : localKeys) {
+                    byte[] workItemBytes = CipherCore.getKeyBytes(workItem);
+
+                    bb = Arrays.copyOf(bb, bb.length + workItemBytes.length);
+                    System.arraycopy(workItemBytes, 0, bb, bb.length - workItemBytes.length,
+                                     workItemBytes.length);
+                }
+                return new SecretKeySpec(bb, "RAW");
+            }
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Perform the HMAC-Extract operation.
+     *
+     * @param inputKey
+     *     the input keying material used for the HKDF-Extract operation.
+     * @param salt
+     *     the salt value used for HKDF-Extract.  If no salt is to be used a {@code null} value
+     *     should be provided.
+     *
+     * @return a byte array containing the pseudorandom key (PRK)
+     *
+     * @throws InvalidKeyException
+     *     if an invalid salt was provided through the {@code HkdfParameterSpec}
+     */
+    protected byte[] hkdfExtract(SecretKey inputKey, byte[] salt) throws InvalidKeyException {
+
+        if (salt == null) {
+            salt = new byte[hmacLen];
+        }
+        hmacObj.init(new SecretKeySpec(salt, "HKDF-Salt"));
+
+        return hmacObj.doFinal(inputKey.getEncoded());
```

Comment:
inputKey can possibly null because consolidateKeyMaterial() returns null when List<SecretKey> keys is null or empty.